db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java

```diff
@@ -1111,15 +1116,28 @@ private void orderUsefulPredicates(Optimizable optTable,
         ArrayList<Predicate> predsForNewInList = null;
     
         if (inlistQualified) {
-          // Only combine multiple IN lists if not using Spark.
-          // Adding extra RDDs and unioning them together hinders performance.
-          if (inlistPreds.size() > 1) {
+          // In DB2 mode, probe keys can be overlapping with eachother.
+          // On HBase the overlapping keys are merged together, but on
+          // mem they aren't causing duplicate rows.  So here we avoid
+          // constructing the overlapping keys on mem.
+          boolean varcharDB2CompatibilityMode = getCompilerContext().getVarcharDB2CompatibilityMode();
+          if (varcharDB2CompatibilityMode && isMemPlatform()) {
+              BinaryRelationalOperatorNode bron =
+                  (BinaryRelationalOperatorNode) usefulPredicates[0].getRelop();
+              boolean isVarChar =
+                  bron.getLeftOperand().getTypeServices().
+                         getTypeName().equals(TypeId.VARCHAR_NAME);
+              if (isVarChar)
+                  usefulCount = 1;
```

Comment:
It feels dangerous to change `usefulCount` to 1 here because `usefulCount` is used not only for multi-probe scan construction but also later to determine start/stop keys. Suppose we have several other predicates together with one in-list on a varchar column so we enter this code path. Setting `usefulCount` to 1 would prevent all other predicates to be scan start/stop keys. Or am I missing anything?

If we don't want to build a `MultiProbe...Scan` for this in-list, is it sufficient to remove it from `inlistPreds` and clean up its states in other data structures (e.g., usefulPredicates[], usefulCount--, isEquality[], etc.) ?

Reply:
This is a good point you bring up.  The intent of this code is to disallow any IN list predicates to be used for multiprobing after the first usefulPredicate which is or type varchar, whether it be in an equality predicate or an in list predicate.  However, usefulCount is not the right position to use.

Just for clarity, let's look at a test case.  Suppose we have this test case:
> create table t (v varchar(14), c char(2));
> insert into t values ('SBVGCCC','  ');
> insert into t values ('SBVGCCC c ','  ');
> insert into t values ('SBVGCCC','A ');
> create index ti on t (v,c);
> select * from t a --splice-properties index=ti
>         where v in ('SBVGCCC    ', 'G')
>         and c in ( ' ', CAST('A' AS CHAR(2)));
We will build 4 probe keys for this.  Two of these are:
Probe Number         Start Key                                    Stop Key
--------------------        -------------                                    ------------
          1                     'SBVGCCC', '  '                           'SBVGCCC       ', '  '
          2                     'SBVGCCC', 'A '                           'SBVGCCC       ', 'A  '

Since the first probe value is a varchar, its start and stop keys must cover a range of values, the start key having zero trailing spaces, and the stop key having the maximum number of trailing spaces to fit the field width.  That means that for all other matching values in between (e.g., 'SBVGCCC ','SBVGCCC  ', 'SBVGCCC   ' ...), for those we also match all values for column c.  The same is true whether we use 'v =' or 'v in'.  However, we could have a third column, i, and have an index:  create index ti2 on t (v,c,i);  with predicates v = 'SBVGCCC    ' and c = 'A ' and i in (1,2,3).  In this case we don't want to disallow use of the column c predicate in the IndexScan because equality predicates are marked as qualifiers, so even if more values than c = 'A ' could initially be scanned, applying the qualifier filters those values out, and this predicate could still be useful as a start key.  Multiprobe IN list predicates don't currently use qualifiers.  I will fix up the code to make the correct usefulCount adjustment and see what other adjustments are appropriate.  Keep in mind, this only affects the mem platform, so isn't something that should affect customers.

Reply:
@ascend1 I've reworked the logic to simply avoid adding the problematic IN list predicates to usefulPredicates, so the adjustment on usefulCount is removed.  I found problem cases on HBase too, so I've removed the distinction between mem platform and non-mem.